pkg/importer/http_data_source.go

```diff
@@ -0,0 +1,280 @@
+)
+
+// HTTPDataSource is the data provider for http(s) endpoints.
+// Sequence of phases:
+// 1a. Info -> Convert (In Info phase the format readers are configured), if the source Reader image is not archived, and no custom CA is used, and can be converted by QEMU-IMG (RAW/QCOW2)
+// 1b. Info -> TransferArchive if the content type is archive
+// 1c. Info -> Transfer in all other cases.
+// 2a. Transfer -> Process if content type is kube virt
+// 2b. Transfer -> Complete if content type is archive (Transfer is called with the target instead of the scratch space). Non block PVCs only.
+// 3. Process -> Convert
+type HTTPDataSource struct {
+	httpReader io.ReadCloser
+	ctx        context.Context
+	cancel     context.CancelFunc
+	cancelLock sync.Mutex
+	// content type expected by the to live on the endpoint.
+	contentType cdiv1.DataVolumeContentType
+	// stack of readers
+	readers *FormatReaders
+	// endpoint the http endpoint to retrieve the data from.
+	endpoint *url.URL
+	// url the url to report to the caller of getURL, could be the endpoint, or a file in scratch space.
+	url *url.URL
+	// true if we are using a custom CA (and thus have to use scratch storage)
+	customCA bool
+}
+
+// NewHTTPDataSource creates a new instance of the http data provider.
+func NewHTTPDataSource(endpoint, accessKey, secKey, certDir string, contentType cdiv1.DataVolumeContentType) (*HTTPDataSource, error) {
+	ep, err := ParseEndpoint(endpoint)
+	if err != nil {
+		return nil, errors.Wrapf(err, fmt.Sprintf("unable to parse endpoint %q", endpoint))
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	httpReader, err := createHTTPReader(ctx, ep, accessKey, secKey, certDir)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	if accessKey != "" && secKey != "" {
+		ep.User = url.UserPassword(accessKey, secKey)
+	}
+	dataProvider := &HTTPDataSource{
+		ctx:         ctx,
+		cancel:      cancel,
+		httpReader:  httpReader,
+		contentType: contentType,
+		endpoint:    ep,
+		customCA:    certDir != "",
+	}
+	// We know this is a counting reader, so no need to check.
+	countingReader := httpReader.(*util.CountingReader)
+	go dataProvider.pollProgress(countingReader, 10*time.Minute, time.Second)
+	return dataProvider, nil
+}
+
+// Info is called to get initial information about the data.
+func (dp *HTTPDataSource) Info() (ProcessingPhase, error) {
+	var err error
+	dp.readers, err = NewFormatReaders(dp.httpReader, dp.contentType)
+	if err != nil {
+		klog.Errorf("Error creating readers: %v", err)
+		return Error, err
+	}
+	// The readers now contain all the information needed to determine if we can stream directly or if we need scratch space to download
+	// the file to, before converting.
+	if dp.readers.Qemu && !dp.readers.Archived && !dp.customCA {
+		// We can pass straight to conversion from the endpoint. No scratch required.
+		dp.url = dp.endpoint
+		return Convert, nil
+	}
+	if dp.contentType == cdiv1.DataVolumeArchive {
+		return TransferTarget, nil
```

Comment:
Can we also return `TransferTarget` for raw image?

Reply:
Done